### PR TITLE
Simplfy handling of save of Nav block uncontrolled inner blocks

### PR DESCRIPTION
**Pull request by getdave**
## What?

Simplifies and centralising the "saving" logic and rendering for uncontrolled inner blocks by defering to the implementation in the parent component (primary `edit.js` of the block).

## Why?

Having a separate usage of `createNavigationMenu` within `uncontroller-inner-blocks.js` was requiring the block to maintain a separate "saving" or "loading" state when creating menus from uncontroller inner blocks (for example converting a Page List into a menu).

This was leading to potential UI bugs and also greatly increasing the complexity of the code within `<UnControlledInnerBlocks>`.

This problem was encountered when working on https://github.com/WordPress/gutenberg/pull/45443.

## How?

With this change we pass in the "create" function and defer all UI handling for saving/loading to the parent component. 

Benefits:

- `<UncontrolledInnerBlocks>` is no longer rendered if a menu is being created avoiding the need to maintain separeate "loading" UI within that component.
- we've been able to remove a lot of logic from the `UncontrolledInnerBlocks` around "loading" states
- we've removed a `setState` call in the parent `edit.js` which simplifies that implementation

## Testing Instructions

The main thing to check is the conversion of inner blocks to sync'd menu.

- Delete all Navigation Menus & Classic Menus - http://localhost:8888/wp-admin/edit.php?post_type=wp_navigation
- New Post
- Add Nav block
- Click on Page List block and click "Edit". Then "Convert".
- See loading state whilst conversion takes place.
- See a _single_ Menu created.

- Delete all Navigation Menus & Classic Menus - http://localhost:8888/wp-admin/edit.php?post_type=wp_navigation
- New Post
- Go to Code view mode and paste in the following:

```
<!-- wp:navigation --><!-- wp:navigation-link {"label":"A new one","type":"post","id":72,"url":"http://localhost:8888/72-2/","kind":"post-type","isTopLevelLink":true} /--><!-- /wp:navigation -->
```

- Switch to Visual Mode.
- Attempt to modify the Navigation.
- See menu created.

### CODE REVIEW

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js:200]: We shouldn't need this state in the parent.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js:693]: Addition of `! isCreatingNavigationMenu` is important here. We shouldn't render the uncontrolled inner blocks if a menu is being created. That way we can defer all the UI logic for a "loading" or "saving" state to the main `edit.js` component.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js:757]: Here's where we pass down the primary creation function.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js:764]: This is now all defer to the centralised logic within the `edit` component which handles creation of menus. Saves maintaining another set of "loading" logic.

File: packages/block-library/src/navigation/edit/unsaved-inner-blocks.js
Comment (getdave) [inline packages/block-library/src/navigation/edit/unsaved-inner-blocks.js:183]: We no longer require this logic as the component will not be rendered when creating a menu.

**Review by draganescu — APPROVED** [review]
Tested and it does what it says on the tin. Also the PR is mostly removing code which is great :)

File: packages/block-library/src/navigation/edit/index.js
Comment (talldan) [reply packages/block-library/src/navigation/edit/index.js:200]: Yeah, agree I don't think it's needed. I think it was only there to support the previous behavior where a menu would be created whenever `UnsavedInnerBlocks` mounted. From memory there where issues where `UnsavedInnerBlocks` would be unmounted and remounted and could cause duplicate menus to be created—the parent block needed a way to guard the menu creation process.

Now this only happens when a user selects/modifies the inner blocks, and I don't think as much guarding is needed. 👍 

I think it's worth giving this a lot of manual testing to be sure, as it's hard to reason about some aspects of the nav block code and also core data resolution.

File: packages/block-library/src/navigation/edit/index.js
Comment (talldan) [inline packages/block-library/src/navigation/edit/index.js]: `navigationMenus` doesn't seem to be used in the effect. It's curious. Guessing it was here because other values are derived from it.

The `useEffect` after this also seems problematic.

```suggestion
```

File: packages/block-library/src/navigation/edit/index.js
Comment (talldan) [inline packages/block-library/src/navigation/edit/index.js:280]: I think `fallbackNavigationMenus` should be memoized (on line 227), otherwise it'll trigger this effect on every render (since it creates a new array each render), and it could be fairly expensive.

Possibly it could also be sorted (lines 248-252) as part of that memoization.

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js]: Great spot 👍

File: packages/block-library/src/navigation/edit/index.js
Comment (getdave) [reply packages/block-library/src/navigation/edit/index.js:280]: > I think fallbackNavigationMenus should be memoized (on line 227), otherwise it'll trigger this effect on every render (since it creates a new array each render), and it could be fairly expensive.

Yes that's correct. Nice spot. I've now memoized it.

> Possibly it could also be sorted (lines 248-252) as part of that memoization.

I agreed with this and so I've added sorting into the memoized value which is recomputed if there are new navigation menus.

**Review by talldan — APPROVED** [review]

**Comment by getdave** [comment]
Waiting on CI tests 🕐

**Pull request merged**
